### PR TITLE
[Doc] Update docstring for some funcs with stable api in ops.py

### PR DESCRIPTION
**Pull request by neozhaoliang**

### CODE REVIEW

File: python/taichi/lang/ops.py [hunk 1]
```diff
@@ -265,94 +373,188 @@ def round(a):  # pylint: disable=redefined-builtin
...
+    so that `log(exp(x)) = x`. The natural logarithm is logarithm in base `e`.
 
     Args:
-        a (Union[:class:`~taichi.lang.expr.Expr`, :class:`~taichi.lang.matrix.Matrix`]): A number or a matrix with elements greater than zero.
+        x (Union[:class:`~taichi.lang.expr.Expr`, :class:`~taichi.lang.matrix.Matrix`]): \
+            Input scalar or matrix.
 
     Returns:
-        The natural logarithm of `a`.
+        The natural logarithm of `x`, element-wise. This is a scalar if `x` is a scalar.
```
Comment (ailzhang) [inline python/taichi/lang/ops.py:519]: Return is the same type as input?

File: python/taichi/lang/ops.py [hunk 2]
```diff
@@ -382,13 +584,39 @@ def logical_not(a):
...
-        A random variable whose type is `dtype`.
+        A random value with type `dtype`.
+
+    Example::
+
+        >>> @ti.kernel
+        >>> def main():
+        >>>     x = ti.random(float)
+        >>>      print(x)
+        0.090257
```
Comment (ailzhang) [inline python/taichi/lang/ops.py:607]: this doesn't look right: you only defined a kernel here but didn't run it? Or maybe you wanted to put them in the comment?

File: python/taichi/lang/ops.py [hunk 3]
```diff
@@ -399,108 +627,218 @@ def random(dtype=float):
...
+                For a scalar and a matrix the scalar will be added element-wise with the matrix.
 
     Returns:
-        sum of `a` and `b`.
+        The sum of `x1` and `x2`, element-wise. This is a scalar if both `x1` and `x2` are scalars.
+
+    Example::
+
+        >>> @ti.kernel
+        >>> def main():
```
Comment (ailzhang) [inline python/taichi/lang/ops.py:644]: nit: better not to name them as `main` :P

File: python/taichi/lang/ops.py [hunk 4]
```diff
@@ -399,108 +627,218 @@ def random(dtype=float):
...
-    """The sub function.
+def sub(x1, x2):
+    """Subtract arguments element-wise.
 
     Args:
-        a (Union[:class:`~taichi.lang.expr.Expr`, :class:`~taichi.lang.matrix.Matrix`]): A number or a matrix.
-        b (Union[:class:`~taichi.lang.expr.Expr`, :class:`~taichi.lang.matrix.Matrix`]): A number or a matrix.
+        x1, x2 (Union[:class:`~taichi.lang.expr.Expr`, :class:`~taichi.lang.matrix.Matrix`]): \
+            The arguments to be subtracted. If both x1, x2 are matrices they must have the \
+                same shape. For a scalar and a matrix the scalar will be applied element-wise \
```
Comment (ailzhang) [inline python/taichi/lang/ops.py:663]: automatically broadcasted to ?

File: python/taichi/lang/ops.py [hunk 5]
```diff
@@ -399,108 +627,218 @@ def random(dtype=float):
...
+    return _binary_operation(_ti_core.expr_sub, _bt_ops_mod.sub, x1, x2)
 
 
 @binary
-def mul(a, b):
-    """The multiply function.
+def mul(x1, x2):
+    """Multiply arguments element-wise.
+
+    Do not confuse this with the matrix multiplication `x1 @ x2` if
```
Comment (ailzhang) [inline python/taichi/lang/ops.py:689]: if -> when, maybe also briefly explain why not to confuse?

File: python/taichi/lang/ops.py [hunk 6]
```diff
@@ -399,108 +627,218 @@ def random(dtype=float):
...
         return _ti_core.expr_sub(a, multiply.ptr)
 
-    return _binary_operation(expr_python_mod, _bt_ops_mod.mod, a, b)
+    return _binary_operation(expr_python_mod, _bt_ops_mod.mod, x1, x2)
 
 
 @binary
-def pow(a, b):  # pylint: disable=W0622
-    """The power function.
+def pow(x1, x2):  # pylint: disable=W0622
```
Comment (ailzhang) [inline python/taichi/lang/ops.py:755]: I personally find that math formulas are more powerful than words here. :P Especially for these ops

**Review by ailzhang — COMMENTED** [review]
Nice work! Thanks a lot!! 
General tips: 
1. No need to mention "return is scalar if input is scalar", but talk more about how input type and output types are related. 
2. List out the math formula whenever it's possible, it's easier to understand for users. 
3. Don't use main as function name, `__main__` is pretty special in python so we want to avoid the confusion here. 
4. Use links to public apis instead of internal `ti.lang....` as much as possible
5. To avoid merge conflict, it might be easier to split the PR to smaller ones in the future ;)

File: python/taichi/_funcs.py [hunk 7]
```diff
@@ -9,44 +9,57 @@
...
 @pyfunc
-def _matrix_transpose(self):
-    """Get the transpose of a matrix.
+def _matrix_transpose(mat):
+    """Permute the first two axes of the matrix.
 
-    Returns:
-        Get the transpose of a matrix.
+    Args:
+        mat (:class:`~taichi.lang.matrix.Matrix`): Input matrix.
```
[the remark this answers is not in the document]
Comment (neozhaoliang) [reply python/taichi/_funcs.py:56]: This formatting is used for sphinx to generate the link:
https://deploy-preview-4328--docsite-preview.netlify.app/api/master/taichi/lang/ops/#taichi.lang.ops.sin
I'm not sure. Does `ti.Matrix` also work` here?

File: python/taichi/lang/ops.py [hunk 8]
```diff
@@ -130,6 +153,34 @@ def cast(obj, dtype):
 
 
 def bit_cast(obj, dtype):
+    """Copy and cast a scalar to a specified data type with its underlying
+    bits preserved. Must be called in taichi scope.
+
+    This function is equivalent to `reinterpret_cast` in C++.
+
+    Args:
+        obj (Union[:class:`~taichi.lang.expr.Expr`]): Input scalar.
```
[the remark this answers is not in the document]
Comment (neozhaoliang) [reply python/taichi/lang/ops.py:162]: Yes. Can we use `ti.types.primitive_types` here instead?

File: python/taichi/lang/ops.py (hunk 2 above)
Comment (neozhaoliang) [reply python/taichi/lang/ops.py:607]: Moved to comments

File: python/taichi/lang/ops.py (hunk 3 above)
Comment (neozhaoliang) [reply python/taichi/lang/ops.py:644]: renamed `main` to `test`

File: python/taichi/lang/ops.py (hunk 4 above)
Comment (neozhaoliang) [reply python/taichi/lang/ops.py:663]: Ok. I didn't quite know what broadcast mean here...

File: python/taichi/lang/ops.py [hunk 9]
```diff
@@ -122,6 +122,29 @@ def wrapped(a, b):
 
 
 def cast(obj, dtype):
+    """Copy and cast a scalar or a matrix to a specified data type.
+    Must be called in Taichi scope.
+
+    Args:
+        obj (Union[:class:`~taichi.lang.expr.Expr`, :class:`~taichi.lang.matrix.Matrix`]): \
```
[the remark this answers is not in the document]
Comment (neozhaoliang) [reply python/taichi/lang/ops.py:129]: Agree. `Expr` is a dead link here. Do you have any suggested candidates?

File: python/taichi/lang/ops.py (hunk 5 above)
Comment (neozhaoliang) [reply python/taichi/lang/ops.py:685]: explanations added

File: python/taichi/lang/ops.py (hunk 6 above)
Comment (neozhaoliang) [reply python/taichi/lang/ops.py:756]: Math formulas may not be exact here. `x^a` could be interpreted as `x xor a`.

Comment (ailzhang) [reply python/taichi/lang/ops.py:756]: I meant "real" math formulas :P https://pytorch.org/docs/stable/generated/torch.pow.html?highlight=pow#torch.pow sphinx should support this in some way

File: python/taichi/_funcs.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ailzhang) [reply python/taichi/_funcs.py:56]: I'm not 100% sure but if you change one, it can be easily verified in the preview link above ;)

**Pull request closed without merging**